### DISCUSSION TITLE
Adding more accurate types for storage settings and improving the docs 💎

### DISCUSSION THREAD

**Discussion by itayperry** [pmndrs/zustand#1955] [Ideas]
Hi everyone,
There's code in the docs shows how to configure the store's storage yourself 🦄:

![image](https://github.com/pmndrs/zustand/assets/37377389/3865cdc9-4661-44ad-b4c6-f13bfe26bf1c)

(btw, this code does not work in TypeScript because `localStorage.getItem(name)` might be `null`)

When running this code the `setItem` type is:
![image](https://github.com/pmndrs/zustand/assets/37377389/200de484-c779-4180-8dee-7d0c25735901)
I believe this should and could be fixed/improved.
### The code shouldn't be `StorageValue<any>`.. because it would definitely be:  
```typescript
StorageValue<BearState>
```

I've written something very similar 🦉:

```typescript
export interface BearState {
  something: string;
  transactions: Set<number>;
  addTransaction: (id: number) => void;
}

export const useBearStore = create<BearState>()(
  persist(
    (set) => ({
      something: "hello",
      transactions: new Set(),
      addTransaction: (x) =>
        set((state) => {
          const transactions = state.transactions;
          transactions.add(x);
          return { transactions: new Set(transactions) };
        }),
    }),
    {
      name: "bears",
      storage: {
        getItem: (name) => {
          const str = localStorage.getItem(name);
          if (!str) return null; // to avoid errors
          return {
            state: {
              ...JSON.parse(str).state,
              transactions: new Set(JSON.parse(str).state.transactions),
            },
          }
        },
        setItem: (name, newValue: StorageValue<BearState>) => {
          const str = JSON.stringify({
            state: {
              ...newValue.state,
              transactions: Array.from(newValue.state.transactions),
            },
          })
          localStorage.setItem(name, str)
        },
        removeItem: (name) => localStorage.removeItem(name),
      },
    }
  )
);
```

The accurate `setItem` type:
![image](https://github.com/pmndrs/zustand/assets/37377389/c967d650-0ed9-467f-98fb-6a0b4d00d558)
This also includes the state methods.. which I don't wanna store in `localStorage` but I understand that you cannot JSON encode a function. 

This is the `getItem` type only when I add `newValue: StorageValue<BearState>` to the `setItem` function, how can I make it more accurate? It shouldn't include the store's methods.
It can return something like: `StorageValue<NoFunctions<BearState>>` or even `StorageValue<Partial<NoFunctions<BearState>>>`, I'll write something down and edit this later.
![image](https://github.com/pmndrs/zustand/assets/37377389/e017d3d2-05e8-407d-b5dc-5b70849ca857)

**Comment by dai-shi** [comment]
Thoughts? @dbritto-dev @sewera

**Reply by dai-shi** [reply]
Closed by #1957

**Comment by itayperry** [comment]
Hi there @dai-shi,
Thank you for merging the PR ✨🥳

Regarding the `getitem` return type, it is `StorageValue<BearState>` but it doesn't really include the methods.

```ts
type NonFunctionPropertyNames = {
  [K in keyof BearState]: BearState[K] extends Function ? never : K;
}[keyof BearState];

type BearStateNonFunctionProperties = Pick<BearState, NonFunctionPropertyNames>;

.
.
.
.

          getItem: (name) => {
            const str = localStorage.getItem(name);
            if (!str) return null;
            const state: BearStateNonFunctionProperties = JSON.parse(str).state;
            return {
              state: {
                ...state,
                transactions: new Set(state.transactions),
              },
            } as StorageValue<BearState>;
          },

```

That's my temporary solution so I could have the `JSON.parse(str).state` as a concrete type. 
I think it needs to return `StorageValue<BearStateNonFunctionProperties>` or  `StorageValue<Partial<BearStateNonFunctionProperties>>`

**Reply by dai-shi** [reply]
While JSON.parse returns `any`, it should be `unknown`. So, the ideal solution is to add a type guard before type casting.

Yeah, you can define what should be the storage value. I think what you mean is `StorageValue<Partial<BearState>>`

**Comment by itayperry** [comment]
>  I think what you mean is `StorageValue<Partial<BearState>>`

But then it means it might include the methods.. which we know for sure will never be there.

What I'm actually saying is that the return type of `getItem` seems wrong

**Reply by dai-shi** [reply]
It's not wrong in a sense `true` is `boolean`.

If you are sure what's your PersistState, it could be `StorageValue<BearStateNonFunctionProperties>` for both getItem and setItem.
The limitation for now is we can't specify two different types for getItem and setItem.

So, in your case, it has to be like this under the limitation:
```ts
        setItem: (name, newValue: StorageValue<BearStateNonFunctionProperties>) => {
```

**Reply by itayperry** [reply]
Oh I see.. well `setItem` will get the methods in `newValue` and `getItem` will never return them.

> The limitation for now is we can't specify two different types for getItem and setItem.

I understand :) 

Anyway, thank you for your help! 
And thank you for going through this discussion and the PR I made, many maintainers ignore these things.. I don't take it for granted 😊

**Reply by dai-shi** [reply]
You are welcome.
Just a nit, but when `setItem` only receives `StorageValue<BearStateNonFunctionProperties>`, you could still pass `StorageValue<BearState>`. (If that doesn't work, it might be somehow improved in the zustand library end.)

**Reply by itayperry** [reply]
@dai-shi 

> but when setItem only receives `StorageValue<BearStateNonFunctionProperties>`, you could still pass `StorageValue<BearState>`

Yes that works :) The code compiles.
It's not perfect but the `getItem` now knows its actual type.

It's like we have to choose whether to get the perfect type in `setItem` or in `getItem`  😅

**Reply by itayperry** [reply]
Btw, I was wrong to write this:
`const state: BearStateNonFunctionProperties = JSON.parse(str).state;`
because even though there are no methods the values themselves (who are not methods) are not all the same, Set and Maps are now Arrays.
Anyway, this specific line was definitely wrong.
